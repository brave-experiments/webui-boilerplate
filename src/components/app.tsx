/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

// React/Redux APIs
import * as React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

// Actions
import * as featureActions from '../actions/featureActions'

// Styles and fonts
import 'emptykit.css'
import '../../assets/style.css'

// Components
import CoreFeature from './feature'

interface Props {
  actions: any
  featureData: any
}

export class App extends React.PureComponent<Props, {}> {
  render () {
    return (
      <div id='featurePage'>
        <CoreFeature featureData={this.props.featureData} actions={this.props.actions} />
      </div>
    )
  }
}

const mapStateToProps = (state: any) => {
  return { featureData: state.featureData }
}

const mapDispatchToProps = (dispatch: any) => {
  return { actions: bindActionCreators(featureActions, dispatch) }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
