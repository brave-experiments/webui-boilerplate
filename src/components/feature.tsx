/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

// React API
import * as React from 'react'

// Types
import { Feature } from '../types/featureTypes'

interface Props {
  actions: any
  featureData: Feature.State
}

interface State {
  addSomebravePeople: Array<string>
}

export default class CoreFeature extends React.PureComponent<Props, State> {
  constructor (props: Props) {
    super(props)
    this.state = {
      addSomebravePeople: [ 'Ross', 'Anthony', 'Serg' ]
    }
  }

  onAddSomebravePeople = (item: string) => {
    this.props.actions.addItem(item)
  }

  clearStorage = () => {
    window.localStorage.clear()
    window.location.reload()
  }

  render () {
    return (
      <div>
        <button onClick={this.clearStorage}>Clear localStorage</button>
        <h2>Brave People</h2>
        <ul>
          {
            this.state.addSomebravePeople.map((item: string, index: number) => {
              return (
                <li key={index}>
                  {item}
                  <button onClick={this.onAddSomebravePeople.bind(this, item)}>(+)</button>
                </li>
              )
            })
          }
        </ul>
        <h2>Amazing Brave people</h2>
        <ol>
          {
            this.props.featureData.bravePeople.map((item: string, index: number) => {
              return <li key={index}>{item}</li>
            })
          }
        </ol>
      </div>
    )
  }
}
