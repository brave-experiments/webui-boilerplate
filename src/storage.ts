/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

// Types
import { Feature } from './types/featureTypes'

// Helpers
import { debounce } from './common/debounce'

// Key to be used by localStorage
const keyName = 'feature-data'

export const defaultState: Feature.State = {
  bravePeople: []
}

export const load = (): Feature.State => {
  const data = window.localStorage.getItem(keyName)
  let state = defaultState
  if (data) {
    try {
      state = JSON.parse(data)
    } catch (error) {
      console.error('Could not parse local storage data: ', error)
    }
  }
  return state
}

export const debouncedSave = debounce<Feature.State>((data: Feature.State) => {
  if (data) {
    window.localStorage.setItem(keyName, JSON.stringify(data))
  }
}, 50)
