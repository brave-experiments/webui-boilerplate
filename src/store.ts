/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

// Redux API
import { createStore } from 'redux'

// Feature core reducer
import rootReducer from './reducers'

export default createStore(rootReducer)
