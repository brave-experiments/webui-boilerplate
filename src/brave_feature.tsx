/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

// React/Redux APIs
import * as React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'

// Main component
import CoreFeature from './components/app'

// Main store
import store from './store'

render((
  <Provider store={store}>
    <CoreFeature />
  </Provider>
  ), document.getElementById('root')
)
