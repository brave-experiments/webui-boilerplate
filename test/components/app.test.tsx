/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

import * as React from 'react'
import { shallow } from 'enzyme'
import { create } from 'react-test-renderer'
import { App } from '../../src/components/app'
import { defaultState } from '../../src/storage'

describe('app dumb component test', () => {
  it('matches the snapshot', () => {
    const component = <App actions={{}} featureData={defaultState} />
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders the component', () => {
    const wrapper = shallow(<App actions={{}} featureData={defaultState} />)
    const assertion = wrapper.find('#featurePage').length
    expect(assertion).toBe(1)
  })
})
