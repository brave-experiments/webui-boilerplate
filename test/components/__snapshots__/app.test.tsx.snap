// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app dumb component test matches the snapshot 1`] = `
<div
  id="featurePage"
>
  <div>
    <button
      onClick={[Function]}
    >
      Clear localStorage
    </button>
    <h2>
      Brave People
    </h2>
    <ul>
      <li>
        Ross
        <button
          onClick={[Function]}
        >
          (+)
        </button>
      </li>
      <li>
        Anthony
        <button
          onClick={[Function]}
        >
          (+)
        </button>
      </li>
      <li>
        Serg
        <button
          onClick={[Function]}
        >
          (+)
        </button>
      </li>
    </ul>
    <h2>
      Amazing Brave people
    </h2>
    <ol />
  </div>
</div>
`;
